### PR TITLE
Propagate status field to cilium network policy

### LINKED ISSUE #1222: Propagate Status field for CiliumNetworkPolicy
A possible status field could look like this:
```
@@ -33,6 +44,80 @@ type CiliumNetworkPolicy struct {

        // Spec is the desired Cilium specific rule specification.
        Spec api.Rule `json:"spec"`
+
+       // Status is the status of the Cilium policy rule
+       Status CiliumNetworkPolicyStatus `json:"status"`
+}
+
+// CiliumNetworkPolicyStatus is the status of a Cilium policy rule
+type CiliumNetworkPolicyStatus struct {
+       // Nodes is the Cilium policy status for each node
+       Nodes map[string]CiliumNetworkPolicyNodeStatus
+
+       // Failing is the number of nodes failing to install this policy
+       Failing int
+}
+
+// CiliumNetworkPolicyNodeStatus is the status of a Cilium policy rule for a
+// specific node
+type CiliumNetworkPolicyNodeStatus struct {
+       // OK is true when the policy has been installed successfully
+       OK bool
+
+       // Message describes the error condition if OK is false
+       Message string
+}
```

### PR DESCRIPTION
**Pull request by aanm**
Fixes #1222

`kubectl get ciliumnetworkpolicies multi-rules -o json`

```
...
    "status": {
        "Nodes": {
            "cilium-k8s-master": {
                "LastSeen": "2017-08-22T20:16:23.471822307-07:00",
                "Message": "\u003cnil\u003e",
                "OK": false
            },
            "cilium-k8s-node-2": {
                "LastSeen": "2017-08-22T20:15:59.540742656-07:00",
                "Message": "\u003cnil\u003e",
                "OK": false
            }
        }
    }

```

### CODE REVIEW

File: pkg/endpoint/policy.go [hunk 1]
```diff
@@ -72,7 +72,7 @@ func (e *Endpoint) evaluateConsumerSource(owner Owner, ctx *policy.SearchContext
 
 	// Skip currently unused IDs
 	if ctx.From == nil || len(ctx.From) == 0 {
-		log.Debugf("[%s] Ignoring unused ID %v", e.PolicyID(), ctx)
+		//log.Debugf("[%s] Ignoring unused ID %v", e.PolicyID(), ctx)
```
Comment (tgraf) [inline pkg/endpoint/policy.go]: Remove this for real and submit as separate PR

File: pkg/k8s/third_party.go [hunk 2]
```diff
@@ -38,6 +40,45 @@ type CiliumNetworkPolicy struct {
...
+// specific node
+type CiliumNetworkPolicyNodeStatus struct {
+	// OK is true when the policy has been installed successfully
+	OK bool
+
+	// Message describes the error condition if OK is false
+	Message string
+
+	// LastSeen contains the last time this status was updated
+	LastSeen time.Time
```
Comment (tgraf) [inline pkg/k8s/third_party.go]: Let's call this `LastUpdated`

File: pkg/k8s/third_party.go [hunk 3]
```diff
@@ -38,6 +40,45 @@ type CiliumNetworkPolicy struct {
...
+}
+
+// CiliumNetworkPolicyNodeStatus is the status of a Cilium policy rule for a
+// specific node
+type CiliumNetworkPolicyNodeStatus struct {
+	// OK is true when the policy has been installed successfully
+	OK bool
+
+	// Message describes the error condition if OK is false
+	Message string
```
Comment (tgraf) [inline pkg/k8s/third_party.go]: Let's call this `Error`

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/k8s/third_party.go (hunk 2 above)
Comment (aanm) [reply pkg/k8s/third_party.go]: Fixed

File: pkg/k8s/third_party.go (hunk 3 above)
Comment (aanm) [reply pkg/k8s/third_party.go]: Fixed

File: pkg/endpoint/policy.go (hunk 1 above)
Comment (aanm) [reply pkg/endpoint/policy.go]: Fixed

File: daemon/k8s_watcher.go [hunk 4]
```diff
@@ -932,11 +937,61 @@ func (d *Daemon) addCiliumNetworkPolicy(obj interface{}) {
...
+		cnpns = k8s.CiliumNetworkPolicyNodeStatus{
+			OK:          true,
+			Error:       "OK",
+			LastUpdated: time.Now(),
+		}
 		log.Infof("Imported CiliumNetworkPolicy %s", rule.Metadata.Name)
 	}
+
+	rule.SetPolicyStatus(nodeaddress.GetName(), cnpns)
+	_, err = crdClient.Update(rule)
```
Comment (tgraf) [inline daemon/k8s_watcher.go]: Move this status update into its own function in `pkg/k8s`

Comment (tgraf) [reply daemon/k8s_watcher.go]: Will the status only work if CRD are in use?

File: daemon/k8s_watcher.go [hunk 5]
```diff
@@ -932,11 +937,61 @@ func (d *Daemon) addCiliumNetworkPolicy(obj interface{}) {
...
+	if err != nil {
+		ns := k8s.ExtractNamespace(&rule.Metadata)
+		name := rule.Metadata.GetObjectMeta().GetName()
+		log.Warningf("k8s: unable to update CNP %s/%s with status: %s, retrying...", ns, name, err)
+		go func(ns, name string, rule *k8s.CiliumNetworkPolicy) {
+			for n := 0; n < 10; n++ {
```
Comment (tgraf) [inline daemon/k8s_watcher.go]: Must be a const in `pkg/k8s/const.go` we should use a single value for all backoff loops so we can adjust globally.

File: daemon/k8s_watcher.go [hunk 6]
```diff
@@ -932,11 +937,61 @@ func (d *Daemon) addCiliumNetworkPolicy(obj interface{}) {
...
+					return
+				}
+				if serverRule.Metadata.UID != rule.Metadata.UID {
+					// Although the policy was found this means it was deleted,
+					// and re-added with the same name.
+					log.Debugf("k8s: rule %s/%s changed while updating node status, stopping retry", ns, name)
+					break
+				}
+				serverRule.SetPolicyStatus(nodeaddress.GetName(), cnpns)
+				_, err = crdClient.Update(serverRule)
```
Comment (tgraf) [inline daemon/k8s_watcher.go]: Doesn't `Update()` give you the latest revision of the resource when it fails? You can use it instead of hoping for a race free get & update.

File: daemon/k8s_watcher.go [hunk 7]
```diff
@@ -963,6 +1018,22 @@ func (d *Daemon) deleteCiliumNetworkPolicy(obj interface{}) {
...
+		log.Warningf("Received unknown object %+v, expected a CiliumNetworkPolicy object", newObj)
+		return
+	}
+	// Since we are updating the status map from all nodes we need to prevent
+	// deletion and addition of all rules in cilium.
+	if oldRule.SpecEquals(newRules) {
+		return
+	}
+
 	// FIXME
```
Comment (tgraf) [inline daemon/k8s_watcher.go]: You can remove the FIXME here now

**Review by tgraf — CHANGES_REQUESTED** [review]

File: daemon/k8s_watcher.go (hunk 4 above)
Comment (aanm) [reply daemon/k8s_watcher.go]: >Will the status only work if CRD are in use?

You got confused because of the variable name? It works for both, but I'll change the variable name to cnpClient

```
	// crdClient is the interface for CRD and TPR
	crdClient k8s.CRDInterface
```

File: daemon/k8s_watcher.go (hunk 6 above)
Comment (aanm) [reply daemon/k8s_watcher.go]: >Doesn't Update() give you the latest revision of the resource when it fails? You can use it instead of hoping for a race free get & update.

It doesn't and it took me a while to figure out that. How can we still get a race with the get & update? I was hopping this would prevent it:
```
if serverRule.Metadata.UID != rule.Metadata.UID {
	// Although the policy was found this means it was deleted,
	// and re-added with the same name.
	log.Debugf("k8s: rule %s/%s changed while updating node status, stopping retry", ns, name)
	break
}
```
but I can modify it to this as well:
```
if serverRule.Metadata.UID != rule.Metadata.UID || !serverRule.SpecEquals(rule)   {
	// Although the policy was found this means it was deleted,
	// and re-added with the same name.
	log.Debugf("k8s: rule %s/%s changed while updating node status, stopping retry", ns, name)
	break
}
```

File: daemon/k8s_watcher.go (hunk 7 above)
Comment (aanm) [reply daemon/k8s_watcher.go]: Done

File: daemon/k8s_watcher.go (hunk 4 above)
Comment (tgraf) [reply daemon/k8s_watcher.go]: Renaming will help. I'm also fine keeping it as-is as CRD will be the only interface soon and TPR will be gone. I was just wondering in general.

File: daemon/k8s_watcher.go (hunk 6 above)
Comment (tgraf) [reply daemon/k8s_watcher.go]: The code is fine but it looked inefficient. It's unexpected that `Update()` does not provide the latest version.

File: daemon/k8s_watcher.go (hunk 5 above)
Comment (aanm) [reply daemon/k8s_watcher.go]: Fixed

File: daemon/k8s_watcher.go (hunk 4 above)
Comment (aanm) [reply daemon/k8s_watcher.go]: Fixed

**Review by tgraf — APPROVED** [review]

**Comment by aanm** [comment]
Finally, this was failing because of the CI tests failures but one of them was caused by me. I've fix it and it has passed.

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
Great work!
